test/test_block_radix_rank.cu

```diff
@@ -0,0 +1,342 @@
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+// Ensure printing of CUDA runtime errors to console
+#include "cub/util_type.cuh"
+#define CUB_STDERR
+
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_radix_rank.cuh>
+#include <cub/block/block_store.cuh>
+#include <cub/block/radix_rank_sort_operations.cuh>
+#include <cub/util_allocator.cuh>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+
+#include "test_util.h"
+#include <stdio.h>
+
+bool g_verbose = false;
+cub::CachingDeviceAllocator g_allocator(true);
+
+template <cub::RadixRankAlgorithm RankAlgorithm,
+          int BlockThreads,
+          int ItemsPerThread,
+          int RadixBits,
+          cub::BlockScanAlgorithm ScanAlgorithm,
+          int Descending,
+          typename Key>
+__launch_bounds__(BlockThreads, 1) __global__ void kernel(Key *d_keys, int *d_ranks)
+{
+  using block_radix_rank = cub::detail::
+    block_radix_rank_t<RankAlgorithm, BlockThreads, RadixBits, Descending, ScanAlgorithm>;
+
+  using storage_t = typename block_radix_rank::TempStorage;
+
+  // Allocate temp storage in shared memory
+  __shared__ storage_t temp_storage;
+
+  // Items per thread
+  Key keys[ItemsPerThread];
+  int ranks[ItemsPerThread];
+
+  constexpr bool uses_warp_striped_arrangement =
+    RankAlgorithm == cub::RadixRankAlgorithm::RADIX_RANK_MATCH ||
+    RankAlgorithm == cub::RadixRankAlgorithm::RADIX_RANK_MATCH_EARLY_COUNTS_ANY ||
+    RankAlgorithm == cub::RadixRankAlgorithm::RADIX_RANK_MATCH_EARLY_COUNTS_ATOMIC_OR;
+
+  if (uses_warp_striped_arrangement)
```

Comment:
Since it's a test that's probably optimized by nvcc in any way, I don't see a reason to deteriorate readability. Is there anything I'm missing?

Reply:
I'm going to merge the PR. If there's a reason to change this, please let me know and I'll update the code in a follow up PR.

Reply:
Some compilers will issue a warning when a constant conditional is evaluated in a regular `if (...)`. We should be using the `if constexpr` macros in public headers for these cases. Since this is just a test, we can fix it later if it becomes an issue.